Build a URL Shortener with React, TypeScript, and Directus/index.md.md

```diff
@@ -0,0 +1,359 @@
+import { useEffect, useState } from "react";
+
+interface ShortLink {
+  clicks: number;
+  date_created?: string;
+  date_updated?: string;
+  id: number;
+  slug: string;
+  sort?: null;
+  url: string;
+  user_created?: string;
+  user_updated?: null;
+}
+
+function App() {
+  return (
+    <>
+      <h1>Link Shortener</h1>
+      <BrowserRouter>
+        <Routes>
+          <Route path="/:slug" element={<LinkRoute />}></Route>
+        </Routes>
+      </BrowserRouter>
+    </>
+  );
+}
+
+function LinkRoute() {
+  const { slug } = useParams();
+  const [slugError, setSlugError] = useState("");
+
+  useEffect(() => {
+    // Perform data fetching based on slug
+    async function fetchShortLink() {
+      try {
+        const data = await directus.request(readItems("short_link"));
+
+        const slug_data = data
+          .map((y) => y)
+          .filter((z) => z.slug.toLowerCase().includes(slug));
+
+        if (!slug_data || slug_data?.length === 0) {
+          setSlugError(`Invalid Slug: Couldn't find the record →→ ${slug}`);
+          throw new Error("Invalid Slug: Couldn't find that record");
+        }
+        const shortLink = slug_data[0] as ShortLink;
+
+        await directus.request(
+          updateItem("short_link", shortLink.id, {
+            clicks: shortLink.clicks + 1,
+          })
+        );
+
+        window.location.assign(`${shortLink.url}`);
+      } catch (error) {
+        console.log(error);
+      }
+    }
+
+    fetchShortLink();
+  }, [slug]);
+
+  return <>{slugError && <h1 style={{ color: "red" }}>{slugError}</h1>}</>;
+}
+
+export default App;
+```
+
+In the preceding code, the single route `//:slug` matches any URL with a slug parameter, and renders the `LinkRoute` component.
+
+The `LinkRoute` component uses the `useParams` hook to get the slug parameter from the URL. It also uses the `useState` hook to store an error message if the slug is invalid.
+
+The `useEffect` hook is used to fetch data from the Directus API via the `fetchShortLink` effect function, which performs the following actions:
+
+> The `ShortLink` interface defines the shape of the short link data returned by the Directus API.
+
+1. It makes a request to the Directus API to read all short links via the `readItems` composable.
+2. The response data is filtered to find the short link that matches the current slug.
+3. If no matching short link is found, it sets an error message and throws an error.
+4. If a matching short link is found, it updates the short links `click` property by `1` via the `updateItem` composable from Directus API.
+5. Finally, the user is redirected to the URL associated with the short link using the `window.location.assign` method.
+
+Finally, if an error occurs during the data fetching process, the error is caught and logged to the console. The `LinkRoute` component also returns the error message if the slug is invalid.
+
+Browser Output:
+
+![Slug URL Output](<Querying Directus.gif>)
```

Comment:
Please can you take this section into smaller, more digestible steps. Please also make use of our diff formatting (check the writing guide) to draw attention to additions/deletions.

Reply:
Done.